atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -205,42 +229,33 @@ private void debugLogStateOfPool() {
 
     private void checkAndUpdateBlacklist() {
         // Check blacklist and re-integrate or continue to wait as necessary
-        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : blacklistedHosts.entrySet()) {
+        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : unresponsiveHostBlacklist.getHosts().entrySet()) {
             long backoffTimeMillis = TimeUnit.SECONDS.toMillis(config.unresponsiveHostBackoffTimeSeconds());
             if (blacklistedEntry.getValue() + backoffTimeMillis < System.currentTimeMillis()) {
                 InetSocketAddress host = blacklistedEntry.getKey();
                 if (isHostHealthy(host)) {
-                    blacklistedHosts.remove(host);
+                    unresponsiveHostBlacklist.removeHost(host);
                     log.error("Added host {} back into the pool after a waiting period and successful health check.", host);
                 }
             }
         }
     }
 
-    private void addToBlacklist(InetSocketAddress badHost) {
-        blacklistedHosts.put(badHost, System.currentTimeMillis());
-    }
-
     private boolean isHostHealthy(InetSocketAddress host) {
-        CassandraClientPoolingContainer testingContainer = null;
         try {
-            testingContainer = new CassandraClientPoolingContainer(host, config);
-            testingContainer.runWithPooledResource(describeRing);
+            currentPools.get(host).runWithPooledResource(describeRing);
```

Comment:
The commit message omitted this, but iterating through the blacklist and removing pools both happen in refreshPool() which is synchronized so I think it's still fine; but I agree that it's probably better to just rewrite this bit. If a pool doesn't exist, then it implies the host is no longer in the config (assuming we preserve the invariant the commit message describes) so we don't need to check its health. So maybe we can just do the check and then do nothing otherwise (which feels scary if you're paranoid, but I think we can maintain this invariant pretty simply and document it).

This does raise the point of "what happens if some thread is shutting down the pool while we're using it"? I think that's actually a problem with normal usage right now. Do we need a R/W lock? Or is there a simpler way to make refreshPool() block until everyone's done using a pool before shutting it down?